cmd/apiserver-watcher/azure_handler.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/golang/glog"
+	"github.com/vishvananda/netlink"
+)
+
+// Prevent hairpin traffic when the apiserver is up
+
+// As per the Azure documentation (https://docs.microsoft.com/en-us/azure/load-balancer/concepts//limitations),
+// if a backend is load-balanced to itself, then the traffic will be dropped.
+//
+// This is because the L3LB does DNAT, so while the outgoing packet has a destination
+// IP of the VIP, the incoming load-balanced packet has a destination IP of the
+// host. That means that it "sees" a syn with the source and destination
+// IPs of itself, and duly replies wit a syn-ack back to itself. However, the client
+// socket expects a syn-ack with a source IP of the VIP, so it drops the packet.
+//
+// The solution is to redirect traffic destined to the lb vip back to ourselves.
+//
+// Example rules using vip 10.0.0.1:
+//
+// -A PREROUTING -m comment --comment "azure LB vip overriding for pods" -j azure-vips
+// -A OUTPUT -m comment --comment "azure LB vip overriding for local clients" -j azure-vips
+// -A azure-vips --dst 10.0.0.1 -j REDIRECT
+
+type azureHandler struct {
+	vip    string
+	ipt    *iptables.IPTables
+	stopCh chan struct{}
+}
+
+var _ handler = &azureHandler{}
+
+func newAzureHandler(vip string, isIPv6 bool) (handler, error) {
+	proto := iptables.ProtocolIPv4
+	if isIPv6 {
+		proto = iptables.ProtocolIPv6
+	}
+
+	ipt, err := iptables.New(
+		iptables.IPFamily(proto),
+		iptables.Timeout(iptablesTimeout),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &azureHandler{
```

Comment:
@aojea do we need to remove stale VIP rules on startup? Looks like the old script might do that:

```
remove_stale() {
        ## find extra iptables rules
        for ipt_vip in $(iptables -w -t nat -S "${CHAIN_NAME}" | awk '$4{print $4}' | awk -F/ '{print $1}'); do
            if [[ ! -v v4vips[${ipt_vip}] ]] || [[ "${v4vips[${ipt_vip}]}" = down ]]; then
                echo removing stale vip "${ipt_vip}" for local clients
                iptables -w -t nat -D "${CHAIN_NAME}" --dst "${ipt_vip}" -j REDIRECT
            fi
        done

```

Reply:
my observation is that when the pod starts it always goes to onfailure state, and whenever you calls that it flush the rules there
```
// deleteRules deletes the iptables rules added for the VIPs
func (az *azureHandler) deleteRules() error {
	glog.Infof("Deleting iptables VIP rules for: %v", az.vip)
	return az.ipt.ClearChain("nat", azureMasqChainName)
}
```

Reply:
Fair enough.